src/farcasterd/syncer_state_machine.rs

```diff
@@ -0,0 +1,327 @@
+use bitcoin::hashes::{hex::ToHex, Hash};
+use farcaster_core::blockchain::{Blockchain, Network};
+
+use crate::{
+    error::Error,
+    event::{Event, StateMachine},
+    rpc::Request,
+    syncerd::{Event as SyncerEvent, SweepAddress, SweepAddressAddendum, Task, TaskId},
+    ServiceId,
+};
+
+use super::runtime::{syncer_up, Runtime};
+
+/// State machine for making a syncer request from and to a client.
+/// State machine automaton:
+/// ```ignore
+///        Start
+///          |
+///    ______|_______
+///   |             |
+///   |             V
+///   |      AwaitingSyncer
+///   |             |
+///   V             V
+/// AwaitingSyncerRequest
+///          |
+///          V
+///         End
+/// ```
+#[derive(Display)]
+pub enum SyncerStateMachine {
+    /// Start state - transitions to AwaitingSyncer or AwaitingSyncerRequest on
+    /// cli request or None on failure.  Transition to AwaitingSyncer triggers
+    /// launch syncer.  Transition to AwaitingSyncerRequest triggers a request
+    /// to the target syncer.
+    #[display("Start")]
+    Start,
+
+    /// AwaitingSyncer state - transitions to AwaitingSyncerRequest once the
+    /// syncer Hello is reiceved. Transition to AwaitingSyncerRequest triggers a
+    /// request to the target syncer.
+    #[display("Awaiting Syncer")]
+    AwaitingSyncer(AwaitingSyncer),
+
+    /// AwaitingSyncerRequest state - transitions to None on success. Transition
+    /// to None triggers a response back to the client.
+    #[display("Awaiting Syncer Request")]
+    AwaitingSyncerRequest(AwaitingSyncerRequest),
+}
+
+pub struct AwaitingSyncer {
+    source: ServiceId,
+    syncer: ServiceId,
+    syncer_task: Task,
+    syncer_task_id: TaskId,
+}
+
+pub struct AwaitingSyncerRequest {
+    syncer_task_id: TaskId,
+    source: ServiceId,
+    syncer: ServiceId,
+}
```

Comment:
just for readability, would be nice to have the `AwaitingSyncer`/`AwaitingSyncerRequest` fields sorted consistently.

Reply:
6b7e3fd59a02e79a24395ebe47706f30ac111483